File: Android/app/src/main/java/com/dreamproject/inwords/presentation/viewScenario/authorisation/AuthorisationViewModel.java
package com.dreamproject.inwords.presentation.viewScenario.authorisation;

import com.dreamproject.inwords.core.util.Event;
import com.dreamproject.inwords.data.dto.UserCredentials;
import com.dreamproject.inwords.domain.interactor.authorisation.AuthorisationInteractor;
import com.dreamproject.inwords.presentation.viewScenario.BasicViewModel;

import java.util.concurrent.TimeUnit;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

public abstract class AuthorisationViewModel extends BasicViewModel {
    private final MutableLiveData<Event<AuthorisationViewState>> authorisationStateLiveData;
    private final MutableLiveData<Event<Boolean>> navigateToLiveData;

    protected AuthorisationInteractor authorisationInteractor;

    protected AuthorisationViewModel(AuthorisationInteractor authorisationInteractor) {
        authorisationStateLiveData = new MutableLiveData<>();
        navigateToLiveData = new MutableLiveData<>();

        this.authorisationInteractor = authorisationInteractor;
    }

    public LiveData<Event<AuthorisationViewState>> getAuthorisationStateLiveData() {
        return authorisationStateLiveData;
    }

    public LiveData<Event<Boolean>> getNavigateToLiveData() {
        return navigateToLiveData;
    }

    public void onNavigateHandler(Observable<Object> clicksObservable) {
        Disposable d = clicksObservable
                .debounce(200, TimeUnit.MILLISECONDS)
                .subscribe(__ -> navigateToLiveData.postValue(new Event<>(true)));
        compositeDisposable.add(d);
    }

    public void onSignHandler(Observable<Object> clicksObservable, Observable<UserCredentials> userCredentialsObservable) {
        Disposable d = clicksObservable
                .debounce(200, TimeUnit.MILLISECONDS)
                .doOnNext(__ -> authorisationStateLiveData.postValue(new Event<>(AuthorisationViewState.loading())))
                .switchMap(__ -> userCredentialsObservable)
                .subscribe(userCredentials ->
                        compositeDisposable.add(performAuthAction(userCredentials)
                                .subscribe(() -> authorisationStateLiveData.postValue(new Event<>(AuthorisationViewState.success())),
                                        t -> authorisationStateLiveData.postValue(new Event<>(AuthorisationViewState.error(t)))))
                );

        compositeDisposable.add(d);
    }

    protected abstract Completable performAuthAction(UserCredentials userCredentials);
}

File: Android/app/src/main/java/ru/inwords/inwords/presentation/view_scenario/authorisation/AuthorisationViewModel.kt
package ru.inwords.inwords.presentation.view_scenario.authorisation

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import ru.inwords.inwords.core.Event
import ru.inwords.inwords.domain.interactor.authorisation.AuthorisationInteractor
import ru.inwords.inwords.presentation.view_scenario.BasicViewModel
import ru.inwords.inwords.profile.data.bean.UserCredentials

abstract class AuthorisationViewModel protected constructor(protected var authorisationInteractor: AuthorisationInteractor) : BasicViewModel() {
    private val authorisationStateLiveData = MutableLiveData<Event<AuthorisationViewState>>()
    private val navigateToLiveData = MutableLiveData<Event<Unit>>()

    val authorisationState: LiveData<Event<AuthorisationViewState>> = authorisationStateLiveData
    val navigateTo: LiveData<Event<Unit>> = navigateToLiveData

    fun onNavigateClicked() {
        navigateToLiveData.postValue(Event(Unit))
    }

    fun onSignClicked(userCredentialsObservable: Observable<UserCredentials>) {
        authorisationStateLiveData.value = Event(AuthorisationViewState.loading())

        userCredentialsObservable
            .flatMapSingle { userCredentials ->
                val validationErrorViewState = validateInput(userCredentials)

                if (validationErrorViewState != null) {
                    Single.just(validationErrorViewState)
                } else {
                    performAuthAction(userCredentials)
                        .andThen(Single.just(AuthorisationViewState.success()))
                        .onErrorResumeNext { Single.just(AuthorisationViewState.error(it)) }
                }
            }
            .subscribe {
                authorisationStateLiveData.postValue(Event(it))
            }
            .autoDispose()
    }

    protected abstract fun performAuthAction(userCredentials: UserCredentials): Completable

    private fun validateInput(userCredentials: UserCredentials): AuthorisationViewState? {
        val emailError = validateEmail(userCredentials)
        val passwordError = validatePassword(userCredentials)

        return when {
            emailError != null && passwordError != null -> {
                AuthorisationViewState.invalidInput(emailError, passwordError)
            }
            emailError != null -> {
                AuthorisationViewState.invalidEmail(emailError)
            }
            passwordError != null -> {
                AuthorisationViewState.invalidPassword(passwordError)
            }

            else -> null
        }
    }

    private fun validateEmail(userCredentials: UserCredentials): String? {
        return if (userCredentials.email.length < 3) {
            "Invalid email"
        } else null
    }

    private fun validatePassword(userCredentials: UserCredentials): String? {
        return if (userCredentials.password.isEmpty()) {
            "Invalid password"
        } else null
    }
}